Framework/NexusGeometry/src/NexusGeometrySave.cpp

```diff
@@ -0,0 +1,1038 @@
+
+  const auto detIds = detInfo.detectorIDs();
+
+  Eigen::Vector3d position =
+      Mantid::Kernel::toVector3d(compInfo.position(index));
+  Eigen::Quaterniond rotation =
+      Mantid::Kernel::toQuaterniond(compInfo.rotation(index));
+
+  std::string dependency = "."; // dependency initialiser
+  std::string name = groupName(compInfo, groupType, index); // group name
+
+  bool locationIsOrigin = isApproxZero(toStdVector(position), PRECISION);
+  bool orientationIsZero = isApproxZero(toStdVector(rotation), PRECISION, true);
+
+  H5::Group childGroup = parentGroup.createGroup(name);
+  writeStrAttribute(childGroup, NX_CLASS, NX_DETECTOR);
+
+  H5::Group transformations =
+      simpleNXSubGroup(childGroup, TRANSFORMATIONS, NX_TRANSFORMATIONS);
+
+  // Orientation is the default first dependency in the chain. first check
+  // translation in component is non-zero, and set dependency to location
+  // if true and write location. Then check if orientation in component is
+  // non-zero, replace dependency with orientation if true. If neither
+  // orientation nor location are non-zero, component is self dependent.
+  if (!locationIsOrigin) {
+    dependency =
+        forwardCompatibility::getObjName(transformations) + "/" + LOCATION;
+    writeLocation(transformations, compInfo, index);
+  }
+  if (!orientationIsZero) {
+    dependency =
+        forwardCompatibility::getObjName(transformations) + "/" + ORIENTATION;
+    writeOrientation(transformations, compInfo, index, locationIsOrigin);
+  }
+
+  H5::StrType dependencyStrType = strTypeOfSize(dependency);
+  writeXYZPixeloffset(childGroup, compInfo, index);
+  writeNXDetectorNumber(childGroup, compInfo, detIds, index);
+
+  writeStrDataset(childGroup, LOCAL_NAME, name);
+  writeStrDataset(childGroup, DEPENDS_ON, dependency);
+}
+
+/*
+ * Function: Saveinstrument
+ * writes the instrument from memory to disk in Nexus format.
+ *
+ * @param compInfo : componentInfo object.
+ * @param fullPath : save destination as full path.
+ * @param reporter : report to progressBase.
+ */
+void saveInstrument(
+    const std::pair<std::unique_ptr<Geometry::ComponentInfo>,
+                    std::unique_ptr<Geometry::DetectorInfo>> &instrPair,
+    const std::string &fullPath, const std::string &rootPath,
+    Kernel::ProgressBase *reporter) {
+
+  const Geometry::ComponentInfo &compInfo = (*instrPair.first);
+  const Geometry::DetectorInfo &detInfo = (*instrPair.second);
+
+  // Exception handling.
+  boost::filesystem::path tmp(fullPath);
+  if (!boost::filesystem::is_directory(tmp.root_directory())) {
+    throw std::invalid_argument(
+        "The path provided for saving the file is invalid: " + fullPath + "\n");
+  }
+  const auto ext = boost::filesystem::path(tmp).extension();
+  if ((ext != ".nxs") && (ext != ".hdf5")) {
+    throw std::invalid_argument("invalid extension for file: " +
+                                ext.generic_string());
+  }
+  if (reporter != nullptr) {
+    reporter->report();
```

Comment:
Reporting not right. You should report for each `NXDetector` written as discussed.

Reply:
done